packet_forwarder/src/lora_pcap.c

```diff
@@ -0,0 +1,355 @@
+/* fix an issue between POSIX and C99 */
+#if __STDC_VERSION__ >= 199901L
+    #define _XOPEN_SOURCE 600
+#else
+    #define _XOPEN_SOURCE 500
+#endif
+
+#include <arpa/inet.h>
+#include <sys/types.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "loragw_hal.h"
+#include "lora_pcap.h"
+#include "cursor/cursor.h"
+
+#define LINKTYPE_LORATAP 270
+#define LW_SYNC 0x34
+
+/*
+ * The PCAP file header. Using clever runtime checking of the
+ * magic number field, this structure can be written to disk
+ * directly without care of the native CPU byte order.
+ */
+struct pcap_header {
+	uint32_t magic_number;
+	uint16_t version_major;
+	uint16_t version_minor;
+	int32_t	 this_zone;
+	uint32_t sigfigs;
+	uint32_t snaplen;
+	uint32_t network;
+};
+
+struct pcap_packet_header_ns {
+	uint32_t tm_s;
+	uint32_t tm_ns;
+	uint32_t packet_length;
+	uint32_t capture_length;
+};
+
+/*
+ * This is the _packed_ representation of the LoRa TAP V1
+ * encoding scheme used by WireShark. Unfortunately, it places
+ * wider fields at misaligned offsets, making it unsafe to fill and
+ * then serialize this structure to disk. As such, it is unused by
+ * this code, but provided here for reference.
+ */
+struct pcap_lora_tap_header {
+	uint8_t  lt_version;
+	uint8_t  lt_padding;
+	uint16_t lt_length;
+	uint32_t channel_frequency_hz;
+	uint8_t  channel_bandwidth;
+	uint8_t  channel_spreading_factor;
+	int8_t   rssi_packet;
+	int8_t   rssi_max;
+	int8_t   rssi_current;
+	int8_t   rssi_snr;
+	uint8_t  sync_word;
+	/* Additions v1 */
+	uint8_t  gw_eui[8];
+	uint32_t timestamp_1mhz; /* !MISALIGNED */
+	uint8_t  flags;
+	uint8_t  coding_rate;
+	uint16_t data_rate;      /* !MISALIGNED */
+	uint8_t  if_channel;
+	uint8_t  rf_chain;
+	uint16_t tag;            /* !MISALIGNED */
+};
+static size_t kLoRaTapHeaderSize = 35;
+
```

Comment:
I like that you use a literal for the size above (I dislike having to do math when reading code), but I always like to add a static assert in these situations:
```suggestion
__Static_assert(sizeof(struct pcap_lora_tap_header) == kLoRaTapHeaderSize, "check LoRa header size");

```

Reply:
I add the static assert suggestion from memory, so please double check syntax/usage before using if you chose to do so.